spec.bs

```diff
@@ -13,3 +13,538 @@ Group: patcg
+
+  Given [=browsing topics types/configuration version=] |configurationVersion|, [=browsing topics types/taxonomy version=] |taxonomyVersion|, and [=browsing topics types/model version=] |modelVersion|, the <dfn for="browsing topics types">version</dfn> is the result of [=string/concatenating=] « |configurationVersion|, |taxonomyVersion|, |modelVersion| » using "<code>:</code>".
+
+  A <dfn for="browsing topics types">user topics state</dfn> is a struct with the following fields and default values:
+  - <dfn for="user topics state">epochs</dfn>: a list of [=epoch=]s, default to an empty list.
+  - <dfn for="user topics state">hmac key</dfn>: 128 bit number, default to 0.
+
+  An <dfn for="browsing topics types">epoch</dfn> is a struct with the following fields:
+  - <dfn for="epoch">taxonomy</dfn>: a list of integers.
+  - <dfn for="epoch">taxonomy version</dfn>: a string.
+  - <dfn for="epoch">model version</dfn>: a string.
+  - <dfn for="epoch">config version</dfn>: a string.
+  - <dfn for="epoch">top 5 topics with caller origins</dfn>: a list of [=topic with caller origins=].
+  - <dfn for="epoch">time</dfn>: a {{DOMHighResTimeStamp}} (from Unix epoch).
+
+  A <dfn for="browsing topics types">topic with caller origins</dfn> is a struct with the following fields:
+  - <dfn for="topic with caller origins">topic id</dfn>: an integer.
+  - <dfn for="topic with caller origins">caller origins</dfn>: a set of [=origins=].
+
+  A <dfn for="browsing topics types">topics history entry</dfn> is a struct with the following fields and default values:
+  - <dfn for="topics history entry">document id</dfn>: an integer, default to 0.
+  - <dfn for="topics history entry">topics calculation input data</dfn>: a string, default to an empty string.
+  - <dfn for="topics history entry">time</dfn>: a {{DOMHighResTimeStamp}} (from Unix epoch).
+  - <dfn for="topics history entry">topics caller origins</dfn>: an ordered set of [=origins=], default to an empty set.
+
+  A <dfn for="browsing topics types">topics caller context</dfn> is a struct with the following fields:
+  - <dfn for="topics caller context">caller origin</dfn>: an [=origin=].
+  - <dfn for="topics caller context">top level context domain</dfn>: a [=domain=].
+  - <dfn for="topics caller context">timestamp</dfn>: a {{DOMHighResTimeStamp}} (from Unix epoch).
+
+  <h2 id="user-agent-associated-state-header">User agent associated state</h2>
+  Each [=user agent=] has an associated [=browsing topics types/user topics state=] <dfn for="user agent">user topics state</dfn> with [=user topics state/epochs=] initially empty, and [=user topics state/hmac key=] initially a randomly generated 128 bit number.
+
+  Each [=user agent=] has an associated <dfn for="user agent">topics history storage</dfn> to store the information about the visited pages that are needed for topics calculation. It is a [=list=] of [=topics history entries=], initially empty.
```

Comment:
I wasn't focused on performance, just on how it would make the spec prose a bit simpler. But yeah, it's fine to leave it as a flat list if you think that reflects the underlying model better.